guardian/gateway
# a few small dx improvements

## What does this change?
This pull request introduces a few shiny new things to our development toolchain:

* **Caching**: With caching to the filesystem enabled, the startup time of the dev server is halved (most of the time). Both Webpack 5 and `babel-loader` caching strategies are enabled.
  * [Line where we're enabling webpack caching](https://github.com/guardian/gateway/pull/790/files#diff-5b9a9fc478ca743b4fd4f484791799fe226cb3a727edf93b1bba3903e5b45a31R12-R18)
  * [Line where we're enabling `babel-loader` caching](https://github.com/guardian/gateway/pull/790/files#diff-5b9a9fc478ca743b4fd4f484791799fe226cb3a727edf93b1bba3903e5b45a31R25-R32)
* **Type checking (for wp watch)**: Added the [fork-ts-checker-webpack-plugin](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin) (catchy name 😉 ) plugin. It's pretty neat. It runs type checking asychronously on a separate process in the background so you can check your types whilst developing without having to run `tsc`. It logs the output with a pretty message sent to stdout.
  * Also adds [fork-ts-checker-notifier-webpack-plugin](https://github.com/johnnyreilly/fork-ts-checker-notifier-webpack-plugin) which opens a notification if the build fails.
  * [Line where we're adding the two plugins](https://github.com/guardian/gateway/pull/790/files#diff-5b9a9fc478ca743b4fd4f484791799fe226cb3a727edf93b1bba3903e5b45a31R56-R64)
* **yarn dev + stmux**: Added a new command that uses [stmux](https://github.com/rse/stmux), a terminal multiplexer based on tmux to launch `yarn watch` and `yarn watch:server` concurrently in a nice two-pane view. Changed readme to mention `yarn dev`
  * [Line where we're using stmux](https://github.com/guardian/gateway/pull/790/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R22)

Please see included screeshots of all of this below.

## How to test
* **Caching**: Run `yarn watch` a few times. Once to build the initial cache and again to observe the benefits of a hopefully improved dev server start time.
* **Type checking**: Run `yarn watch` on the project and check that the plugin passes the type check. Add a deliberate type error and check back at the console to see if it showed up.
* **yarn dev + stmux**: Run `yarn:dev` and check that both the webpack watch command and nodemon server command are running in their respective panes. To quit, use the mouse to select each pane and ctrl-c out of both. stmux should then close.

## How can we measure success?
✨ _**Developer happiness should increase**_ ✨
(review at a future engineering session?)

## Have we considered potential risks?
Some things that came to mind and my thoughts:

* Changes to the development webpack config could maybe cause discrepancies with the main production build?
  * Seems unlikely to me as the only substatial changes are the enabling of caching and the inclusion of the two new webpack plugins.
* Adds three new **dev-dependencies**:
  * **fork-ts-checker-webpack-plugin**, **fork-ts-checker-notifier-webpack-plugin**, both widely used and trusted webpack plugins.
  * **stmux**, only usage is as a means to display two commands side-by-side.

## Images

### stmux multiplexer
![Screenshot 2021-08-08 at 22 28 43](https://user-images.githubusercontent.com/1771189/128646268-721723df-73a8-44f0-9ae7-be730fd2753a.png)

### Before/After caching start times

#### Before
![Screenshot 2021-08-08 at 22 27 53](https://user-images.githubusercontent.com/1771189/128646249-256c4b3c-dd79-4ad6-8016-c57ae76bfd1b.png)

#### After
![Screenshot 2021-08-08 at 20 43 26](https://user-images.githubusercontent.com/1771189/128646203-ef4b60c1-b9f0-4d39-9b2e-c39867f260db.png)

### fork-ts-checker-webpack-plugin

#### Passing
![Screenshot 2021-08-08 at 22 27 53](https://user-images.githubusercontent.com/1771189/128646249-256c4b3c-dd79-4ad6-8016-c57ae76bfd1b.png)

#### Failing
![Screenshot 2021-08-08 at 20 44 31](https://user-images.githubusercontent.com/1771189/128646177-c08d648b-6ddd-4aeb-a78d-7d80e11b7842.png)


## Accessibility
not applic. here
